### DISCUSSION TITLE
How does this tool extract a webpage's main content?

### DISCUSSION THREAD

**Discussion by bytrangle** [webclipper/web-clipper#966] [Q&A]
I can see that this tool uses an extension named [Readability](https://github.com/webclipper/web-clipper/blob/master/src/extensions/extensions/readability.ts) to "intelligently extract webpage main content".

It is intelligent indeed, and I'm very pleased with the quality of the extraction. I'd like to learn more about it, but I don't understand what the codes do.

```
export default new TextExtension(
  {
    name: 'Readability',
    icon: 'copy',
    version: '0.0.1',
    description: 'Intelligent extraction of webpage main content.',
    i18nManifest: {
      'zh-CN': { name: '智能提取', description: '智能提取当前页面元素' },
    },
  },
  {
    run: async context => {
      const { turndown, document, Readability, $ } = context;
      let documentClone = document.cloneNode(true);
      $(documentClone)
        .find('#skPlayer')
        .remove();
      let article = new Readability(documentClone, {
        keepClasses: true,
      }).parse();
      return turndown.turndown(article.content);
    },
  }
);
```

I guess the `Readability` property here comes from Mozilla's [Readability library](https://github.com/mozilla/readability). But I don't see this library in `package.json`, nor is it imported anywhere in the project. Can anyone explain how this magic happens?

**Comment by DiamondYuan** [comment]
i fork the readability https://github.com/webclipper/readability

**Comment by DiamondYuan** [comment]
And i inject readability to the extension context.

https://github.com/webclipper/web-clipper/blob/eebc4a2ecf34fda83d5c32620fd68b1c52412f70/src/service/contentScript/browser/contentScript/contentScript.ts#L100
